File: day37/swapPairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        res = ListNode(0)
        res.next = head
        cur = res
        while cur.next and cur.next.next:
            first = cur.next
            second = cur.next.next

            # swap
            cur.next = second
            first.next = second.next
            second.next = first

            cur = first
        return res.next

File: day37/swapPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode res = new ListNode(-1);

        res.next = head;
        ListNode cur = res;

        while (cur.next != null && cur.next.next != null) {
            ListNode first = cur.next;
            ListNode second = cur.next.next;

            cur.next = second;
            first.next = second.next;
            second.next = first;

            cur = first;
        }

        return res.next;
    }
}
